hw2_1/HW_1.py

```diff
@@ -0,0 +1,713 @@
+# + Должен иметь метод `show_last_message`, выводящий на экран информацию о последнем сообщении
+# + Должен иметь метод `get_history_from_time_period`, который принимает два опциональных аргумента (даты с которой и по какую мы ищем сообщения и выдаём их). Метод также должен возвращать объект типа `Chat`
+# + Должен иметь метод `show_chat`, выводящий на экран все сообщения (каждое сообщение в таком же виде как и `show_last_message`, но с разделителем между ними)
+# + Должен иметь метод `recieve`, который будет принимать сообщение и добавлять его в чат
+# 
+# **Message**:
+# + Должен иметь три обязательных атрибута
+#     + `text` - текст сообщения
+#     + `datetime` - дата и время сообщения (встроенный модуль datetime вам в помощь). Важно! Это должна быть не дата создания сообщения, а дата его попадания в чат! 
+#     + `user` - информация о пользователе, который оставил сообщение (какой тип данных использовать здесь, разберётесь сами)
+# + Должен иметь метод `show`, который печатает или возвращает информацию о сообщении с необходимой информацией (дата, время, юзер, текст)
+# + Должен иметь метод `send`, который будет отправлять сообщение в чат
+# 
+# **User**:
+# + Класс с информацией о юзере, наполнение для этого класса придумайте сами
+# 
+# Напишите несколько примеров использования кода, которое показывает взаимодействие между объектами.
+# 
+# В тексте задания намерено не указано, какие аргументы должны принимать методы, пускай вам в этом поможет здравый смысл)
+# 
+# В этом задании не стоит флексить всякими продвинутыми штуками, для этого есть последующие
+# 
+# В этом задании можно использовать только модуль `datetime`
+
+# In[1]:
+
+
+from datetime import datetime
+from random import choice   # for User
+
+
+# In[2]:
+
+
+class User:
+    def __init__(self, name='I'):
+        self.name = name 
+
+    def online_offline(self):
+        return choice(['online', 'offline'])
+
+
+# In[3]:
+
+
+class Message:
+   def __init__(self, text, user, datetime = 'Draft'):   #maybe as draft or as planning message 
+       self.text = text
+       self.time = datetime
+       if isinstance(user, User):                        #i think its better then convert into User
+           self.user = user.name
+       else:
+           self.user = User('Incognito').name
+   
+   def show(self):
+       print(f' Message text: {self.text} \n Time: {self.time} \n Sender: {self.user} \n')
+   
+   def send(self, chat):
+       if not isinstance(self.time, datetime):
+           self.time = datetime.now()
+           
+       chat.chat_history.insert(0, self)
+
+
+# In[4]:
+
+
+class Chat:
+    def __init__(self, chat_history = []):
+        self.chat_history = chat_history
+        
+    def show_last_message(self):
+        type(Message)(self.chat_history[0]).show(self.chat_history[0])
+    
+    def get_history_from_time_period(self, start = datetime.now, stop = datetime.min):
```

Comment:
В таких случаях обычно делают значениями по-умолчанию `None`, а в коде функции уже как-то обрабатывают
```suggestion
    def get_history_from_time_period(self, start=None, stop=None):
```